### PR TITLE
[flang][OpenMP] Add semantic check for declare target

### PR DESCRIPTION
**Pull request by shraiysh**
This patch adds the following check from OpenMP 5.2.

```
If the directive has a clause, it must contain at least one enter clause
or at least one link clause.
```

Also added a warning for the deprication of `TO` clause on `DECLARE TARGET` construct.

```
The clause-name to may be used as a synonym for the clause-name enter.
This use has been deprecated.
```

Last attempt to add this caused issues in buildbots - #71861. It should pass now.

### CODE REVIEW

**Comment by shraiysh** [comment]
One thing I wanted to point out was that with this patch clang will accept `#pragma omp declare target enter(x)` when compiled with version 5.2. I don't think clang is handling it accurately yet though. Is there a way to report "unimplemented" errors in clang? Or is it okay this way?

File: flang/lib/Lower/OpenMP.cpp
Comment (skatrak) [inline flang/lib/Lower/OpenMP.cpp:1861]: Nit: Update comment

File: flang/lib/Lower/OpenMP.cpp
Comment (skatrak) [inline flang/lib/Lower/OpenMP.cpp:593]: Nit: It would be good to keep the alphabetic order of declarations here and definitions below, so I'd add this between `processDeviceType` and `processFinal`.

**Review by skatrak — COMMENTED** [review]
Thank you Shraiysh, I just have a couple of nits.

**Review by agozillon — APPROVED** [review]
LGTM, aside from Sergio's Nits! 

Might be worth asking someone from the Clang team to have a look at the Clang changes if you're a little unsure on them, but if you're happy with them can disregard doing that!

**Comment by shraiysh** [comment]
> Might be worth asking someone from the Clang team to have a look at the Clang changes if you're a little unsure on them, but if you're happy with them can disregard doing that!

Thank you for the review @agozillon @skatrak. I am not sure who I should request the review from, but I have requested a review from the codeowner of OpenMP conformance. clang:openmp is also tagged as a topic in this patch, so I will wait for 1-2 days before merging this if someone from clang/openmp team has any concerns with this.

File: flang/lib/Lower/OpenMP.cpp
Comment (shraiysh) [reply flang/lib/Lower/OpenMP.cpp]: Thanks for the review @skatrak. I moved it between `processDepend` and `processIf` because `enter` is a repeatable clause.

**Pull request merged**
